### PR TITLE
Add gata policy

### PR DESCRIPTION
**Pull request by Ashutosh-Adhikari**

### CODE REVIEW

File: maml_rl/samplers/multi_task_sampler.py [hunk 1]
```diff
@@ -304,15 +304,25 @@ def create_episodes(self,
...
+                # obs_strings = [item + " <sep> " + a for item, a in zip(obs_strings, chosen_actions)]
+                # value, chosen_actions, meta_prev_h, action_log_probs, chosen_indices, _, prev_h, prev_c = agent.act(agent.rollouts.ro_observation_list[step], agent.rollouts.ro_current_triplets[step], agent.rollouts.ro_action_candidate_list[step], meta_dones.unsqueeze(1), meta_torch_step_rewards.unsqueeze(1), meta_prev_h)
+                # chosen_actions = [(action if not done else "restart") for done, action in zip(dones, chosen_actions)]
+                # chosen_actions_before_parsing = [(item[idx] if not done else "*restart*") for item, idx, done in zip(infos["admissible_commands"], chosen_indices, dones)]
+                
+                ######
+
                 # TODO:
-                # observations_tensor = torch.from_numpy(observations)
+                # observations_tensor = torch.from_numpy(observations) ## Do we realy want numpy? If so, i will need to demarcate inside the agent.act and essentialy write the function explicitly write here--easy
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py]: Numpy is not needed. That's the old/original code from Tristan.

File: maml_rl/samplers/multi_task_sampler.py [hunk 2]
```diff
@@ -304,15 +304,25 @@ def create_episodes(self,
                                     normalize=True)
         return episodes
 
-    def sample_trajectories(self, params=None):
+    def sample_trajectories(self, params=None): # need to pass Agent() to the class?
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py]: maybe `policy` can be the `agent`.

**Review by MarcCote — CHANGES_REQUESTED** [review]
Yes, that's what I had in mind for the preprocessing stuff.

File: maml_rl/samplers/multi_task_sampler.py (hunk 1 above)
Comment (Ashutosh-Adhikari) [reply maml_rl/samplers/multi_task_sampler.py]: cool

File: maml_rl/samplers/multi_task_sampler.py (hunk 2 above)
Comment (Ashutosh-Adhikari) [reply maml_rl/samplers/multi_task_sampler.py]: yes

**Comment by Ashutosh-Adhikari** [comment]
Alright i will uncomment and push

**Comment by Ashutosh-Adhikari** [comment]
Next step, make `update_params()` and `reinforce_loss()` compatible with `GATA`

File: maml_rl/samplers/multi_task_sampler.py [hunk 3]
```diff
@@ -45,6 +45,10 @@ class MultiTaskSampler(Sampler):
         The policy network for sampling. Note that the policy network is an
         instance of `torch.nn.Module` that takes observations as input and
         returns a distribution (typically `Normal` or `Categorical`).
+    agent : `TWr.agent.Agent`
+        The GATA agent with TWr/model.KG_Manipulation for sampling. Can we used for
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py:61]: "Can we" -> "Can be"

File: maml_rl/samplers/multi_task_sampler.py [hunk 4]
```diff
@@ -308,30 +312,46 @@ def sample_trajectories(self, params=None): # need to pass Agent() to the class?
...
-                actions = ["look"] * self.batch_size
+                # actions = ["look"] * self.batch_size
+                actions = chosen_actions_before_parsing.cpu().numpy()
 
                 new_observations, rewards, _, infos = self.envs.step(actions)
                 batch_ids = infos['batch_ids']
                 yield (observations, actions, rewards, batch_ids)
                 observations = new_observations
+                prev_actions = chosen_actions_before_parsing
+                ## what about the dones? shouldn't env.step return it? do we have batch_ids?
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py]: `infos` here is not the one you are used to. It's a dict containing two keys: `batch_ids` and `infos`. This is why I do
`observations = [info["feedback"] for info in infos["infos"]]`

File: maml_rl/samplers/multi_task_sampler.py [hunk 5]
```diff
@@ -308,30 +312,46 @@ def sample_trajectories(self, params=None): # need to pass Agent() to the class?
...
+                prev_game_facts.append(set())
+                prev_step_dones.append(0.0)
+                prev_scores.append(0.0)
+            ####
+            # Don't need Rl^2 here but just in case
+            meta_dones = to_pt(np.zeros(self.batch_size), enable_cuda=self.agent.use_cuda, type='float')
+            meta_torch_step_rewards = to_pt(np.zeros(self.batch_size), enable_cuda=self.agent.use_cuda, type='float')
+            meta_prev_h = to_pt(np.zeros((1, self.batch_size, self.agent.policy_net.block_hidden_dim)), enable_cuda=self.agent.use_cuda, type='float')
+            ####
+            observations = [info["feedback"] for info in infos["infos"]]
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py]: Note here that `infos` is custom dict that Tristan envs is returning which contains two keys: `batch_ids` and `infos`. The latter is the one containing the infos from TW. However, it is a bit different from what we have in TW, where instead of `Dict[str, List[Any]]` is `List[Dict[str, Any]]`. This means we need to do some conversion here to be compatible with our TWr functions.

You can use this following function to do the conversion:
```python
def _list_of_dicts_to_dict_of_lists(list_: List[Dict]) -> Dict[str, List]:
    # Convert List[Dict] to Dict[List]
    keys = set(key for dict_ in list_ for key in dict_)
    return {key: [dict_.get(key) for dict_ in list_] for key in keys}
```

File: maml_rl/samplers/multi_task_sampler.py [hunk 6]
```diff
@@ -308,30 +312,46 @@ def sample_trajectories(self, params=None): # need to pass Agent() to the class?
...
-
-                ######
-                ## Preprocess
-                # obs_strings, current_triplets, acl, _, current_game_facts = agent.get_game_information_at_certain_step(obs, infos, prev_actions='tw-restart(?)', prev_facts=None)
-                # obs_strings = [item + " <sep> " + a for item, a in zip(obs_strings, chosen_actions)]
-                # value, chosen_actions, meta_prev_h, action_log_probs, chosen_indices, _, prev_h, prev_c = agent.act(agent.rollouts.ro_observation_list[step], agent.rollouts.ro_current_triplets[step], agent.rollouts.ro_action_candidate_list[step], meta_dones.unsqueeze(1), meta_torch_step_rewards.unsqueeze(1), meta_prev_h)
-                # chosen_actions = [(action if not done else "restart") for done, action in zip(dones, chosen_actions)]
-                # chosen_actions_before_parsing = [(item[idx] if not done else "*restart*") for item, idx, done in zip(infos["admissible_commands"], chosen_indices, dones)]
+                observation_strings, current_triplets, action_candidate_list, _, current_game_facts = self.agent.get_game_information_at_certain_step(observations, infos, prev_actions=chosen_actions, prev_facts=None)
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py]: Here `get_game_information_at_certain_step` is expected `infos: Dict[str, List[Any]]`.

File: train.py [hunk 7]
```diff
@@ -35,25 +52,28 @@ def main(args):
...
+    # policy = get_policy_for_env(env,
+    #                            hidden_sizes=config['hidden-sizes'],
+    #                            nonlinearity=config['nonlinearity'])
+    # --replaced by agent
+
+    agent.share_memory() #policy.share_memory()
 
     # Baseline
-    baseline = LinearFeatureBaseline(get_input_size(env))
+    baseline = LinearFeatureBaseline(get_input_size(env)) ## input_size can agent.block_hidden_dim
```
Comment (MarcCote) [inline train.py:71]: This will need to be changed accordingly to take as input the same type of observation as our agent.

**Review by MarcCote — CHANGES_REQUESTED** [review]
The file `train.py` should be in the TWr repo, not maml-rl.

File: train.py (hunk 7 above)
Comment (Ashutosh-Adhikari) [reply train.py:71]: True.

File: maml_rl/samplers/sampler.py [hunk 8]
```diff
@@ -1,8 +1,8 @@
 import gym
 
-def make_env(env_name, env_kwargs={}, seed=None):
+def make_env(env_name, seed=None): #def make_env(env_name, env_kwargs={}, seed=None): - we don't have kwargs for tw-env
     def _make_env():
-        env = gym.make(env_name, **env_kwargs)
+        env = gym.make(env_name) #env = gym.make(env_name, **env_kwargs) -- we don't have kwargs for tw-env
```
Comment (MarcCote) [inline maml_rl/samplers/sampler.py]: I don't think that should fail with tw-env. It'd be better to leave it there (i.e. reducing the changes).

**Review by MarcCote — CHANGES_REQUESTED** [review]
`train.py`should be in TWr right?

File: maml_rl/samplers/multi_task_sampler.py [hunk 9]
```diff
@@ -304,35 +328,67 @@ def create_episodes(self,
...
+                # observations_tensor = torch.from_numpy(observations) ## Do we realy want numpy? If so, i will need to demarcate inside the agent.act and essentialy write the function explicitly write here--easy
                 # pi = self.policy(observations_tensor, params=params)
                 # actions_tensor = pi.sample()
                 # actions = actions_tensor.cpu().numpy()
-                actions = ["look"] * self.batch_size
+                # actions = ["look"] * self.batch_size
+                #actions = chosen_actions_before_parsing.cpu().numpy()
 
-                new_observations, rewards, _, infos = self.envs.step(actions)
+                new_observations, rewards, dones, infos = self.envs.step(chosen_actions_before_parsing)
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py:378]: Here, `new_observations` is a numpy array of zeros. You should use the observations obtained from `infos` (like done above).

Comment (MarcCote) [reply maml_rl/samplers/multi_task_sampler.py:378]: For reference: 
https://github.com/Ashutosh-Adhikari/pytorch-maml-rl/blob/tw-maml/maml_rl/envs/textworld.py#L139

Comment (Ashutosh-Adhikari) [reply maml_rl/samplers/multi_task_sampler.py:378]: i see

Comment (Ashutosh-Adhikari) [reply maml_rl/samplers/multi_task_sampler.py:378]: actually it's been overridden anyway

**Comment by Ashutosh-Adhikari** [comment]
TODO : As @MarcCote said, remove dependency from `layers.py` and `generic.py`

**Comment by Ashutosh-Adhikari** [comment]
@MarcCote , so what I am doing essentially in the commit above is padding once we get the observations rather than padding the input. 
Question : padding the input would require efforts from the env side right?

File: maml_rl/baseline.py [hunk 10]
```diff
@@ -55,19 +55,27 @@ def _feature(self, episodes):
...
+        max_ep_length = len(episodes)
         ep_h_og = []
         ep_obs_mask = []
         ep_h_go = []
         ep_node_mask = []
         for i in range(episodes.batch_size):
             h_og, obs_mask, h_go, node_mask = self.agent.encode(ep_obs_str[i], ep_triplets[i], use_model="policy")
             #ep_h_og.append(h_og.unsqueeze(0))
+
+            h_go = F.pad(h_go, (0, 0, 0, 0, 0, max_ep_length-ep_lengths[i])) # hardcoded not only for graph input--but for padding too--suggestions?
```
Comment (Ashutosh-Adhikari) [inline maml_rl/baseline.py:62]: don't like this hardcoding. essentially i _need_ to know shape of the tensors to pad it.

**Comment by MarcCote** [comment]
To address your question about padding from the env side, yes it would be hard because it's only aware of a single observation. The batching process is done in Gym.

File: maml_rl/baseline.py [hunk 11]
```diff
@@ -1,9 +1,11 @@
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
-
+from debug import ForkedPdb as fpdb
```
Comment (MarcCote) [inline maml_rl/baseline.py]: `debug.py` doesn't exist in this library.

File: maml_rl/baseline.py [hunk 12]
```diff
@@ -1,9 +1,11 @@
...
 from collections import OrderedDict
 
+from layers import masked_mean
```
Comment (MarcCote) [inline maml_rl/baseline.py]: `layers.py` doesn't exist in this library.

File: maml_rl/baseline.py [hunk 13]
```diff
@@ -1,9 +1,11 @@
...
+
 class LinearFeatureBaseline(nn.Module):
```
Comment (MarcCote) [inline maml_rl/baseline.py:9]: I don't think we should be modifying this LinearFeatureBaseline directly. Instead, we should create our own Baseline class in TWr and use it with maml-rl (like we are doing with the policy net, i.e. we defined the model's code in TWr).

File: maml_rl/metalearners/maml_trpo.py [hunk 14]
```diff
@@ -94,8 +109,86 @@ def _product(vector, retain_graph=True):
 
         with torch.set_grad_enabled(old_pi is None):
             valid_episodes = await valid_futures
-            pi = self.policy(valid_episodes.observations, params=params)
-
+            if params is not None: # can make a function
```
Comment (MarcCote) [inline maml_rl/metalearners/maml_trpo.py:116]: What's the purpose of all this below?

File: maml_rl/metalearners/maml_trpo.py [hunk 15]
```diff
@@ -131,8 +224,18 @@ def step(self,
...
+        non_none_grads = []
+        non_none_indices = []
+        for i in range(len(grads)):
+            elem = grads[i]
+            if elem is not None:
+                non_none_indices.append(i)
+                non_none_grads.append(elem.contiguous())
+        grads = tuple(non_none_grads)
+        #print("Type of grads")
+        #print(grads)
```
Comment (MarcCote) [inline maml_rl/metalearners/maml_trpo.py:242]: Is this because we have parameters that we don't update at all?

File: maml_rl/samplers/multi_task_sampler.py [hunk 16]
```diff
@@ -11,15 +15,21 @@
 from maml_rl.envs.utils.sync_vector_env import SyncVectorEnv
 from maml_rl.episode import BatchEpisodes
 from maml_rl.utils.reinforcement_learning import reinforce_loss
-
+from generic import to_pt
```
Comment (MarcCote) [inline maml_rl/samplers/multi_task_sampler.py]: generic.py not part of this module

**Review by MarcCote — CHANGES_REQUESTED** [review]
Also, please remove any commented lines, debugging code. It is very hard to review real changes when I have to dig for them since Github interface just highlights all new lines in green!

File: maml_rl/baseline.py (hunk 13 above)
Comment (Ashutosh-Adhikari) [reply maml_rl/baseline.py:9]: This is our modified baseline.

File: maml_rl/metalearners/maml_trpo.py (hunk 15 above)
Comment (Ashutosh-Adhikari) [reply maml_rl/metalearners/maml_trpo.py:242]: yes!

File: maml_rl/samplers/multi_task_sampler.py [hunk 17]
```diff
@@ -294,39 +323,67 @@ def create_episodes(self,
...
         episodes.log('duration', time.time() - t0)
-
-        self.baseline.fit(episodes)
-        episodes.compute_advantages(self.baseline,
+        self.baseline.fit(episodes, self.agent)
+        episodes.compute_advantages(self.baseline, self.agent,
                                     gae_lambda=gae_lambda,
                                     normalize=True)
+        if params is not None:
+            self.agent.policy_net.load_state_dict(old_params) # hope there is no race condition here! NOTE: strict=True
```
Comment (Ashutosh-Adhikari) [inline maml_rl/samplers/multi_task_sampler.py:337]: @MarcCote i think this is fundamentally different than tristan's corresponding code :
1) He is using `params` on the fly for the policy net.
2) We need `params` for baseline. 

Comments requested on this.

Comment (MarcCote) [reply maml_rl/samplers/multi_task_sampler.py:337]: Right, so having the strings already embedded, you don't need the policy's params.

File: maml_rl/baseline.py [hunk 18]
```diff
@@ -42,15 +46,45 @@ def _feature(self, episodes):
             ones
         ], dim=2)
 
-    def fit(self, episodes):
+    def _tw_feature(self, episodes, agent):
```
Comment (Ashutosh-Adhikari) [inline maml_rl/baseline.py:49]: Is this _feature extraction of fitting valid?
